ms2deepscore/spectrum_pair_selection.py

```diff
@@ -0,0 +1,278 @@
+    fingerprints: np.ndarray,
+    selection_bins: np.ndarray = np.array([(x/10, x/10 + 0.1) for x in range(0, 10)]),
+    max_pairs_per_bin: int = 20,
+    include_diagonal: bool = True,
+    fix_global_bias: bool = True,
+) -> np.ndarray:
+    """Returns matrix of jaccard indices between all-vs-all vectors of references
+    and queries.
+
+    Parameters
+    ----------
+    fingerprints
+        Fingerprint vectors as 2D numpy array.
+    selection_bins
+        List of tuples with upper and lower bound for score bins.
+        The goal is to pick equal numbers of pairs for each score bin.
+        Sidenote: bins do not have to be of equal size, nor do they have to cover the entire
+        range of the used scores.
+    max_pairs_per_bin
+        Specifies the desired maximum number of pairs to be added for each score bin.
+    include_diagonal
+        Set to False if pairs with two equal compounds/fingerprints should be excluded.
+    fix_global_bias
+        Default is True in which case the function aims to get the same amount of pairs for
+        each bin globally. This means it add more than max_pairs_par_bin for some bins and/or
+        some compounds to compensate for lack of such scores in other compounds.
+
+    Returns
+    -------
+    scores
+        Sparse array (List of lists) with cherrypicked scores.
+    """
+    # pylint: disable=too-many-locals
+    size = fingerprints.shape[0]
+    scores_data = []
+    scores_i = []
+    scores_j = []
+    # keep track of total bias across bins
+    max_pairs_global = len(selection_bins) * [max_pairs_per_bin]
+    for i in range(size):
+        scores_row = np.zeros(size)
+        for j in range(size):
+            if i == j and not include_diagonal:
+                continue
+            scores_row[j] = jaccard_index(fingerprints[i, :], fingerprints[j, :])
+
+        # Cherrypicking
+        for bin_number, selection_bin in enumerate(selection_bins):
+            # Indices of scores within the current bin
+            idx = np.where((scores_row > selection_bin[0]) & (scores_row <= selection_bin[1]))[0]
+
+            # Randomly select up to max_pairs_per_bin scores within the bin
+            #if len(idx) > 0:
+            np.random.shuffle(idx)
+            if fix_global_bias:
+                idx_selected = idx[:max_pairs_global[bin_number]]
+                max_pairs_global[bin_number] += max_pairs_per_bin
+                max_pairs_global[bin_number] -= len(idx_selected)  # only remove actually found pairs
+            else:
+                idx_selected = idx[:max_pairs_per_bin]
+            scores_data.extend(scores_row[idx_selected])
+            scores_i.extend(len(idx_selected) * [i])
+            scores_j.extend(list(idx_selected))
+    # Output biases in scores
+    print("Counted cases of fewer than max_pairs_per_bin usage:")
+    max_pairs_global = [x - max_pairs_per_bin for x in max_pairs_global]
+    print(max_pairs_global)
```

Comment:
I know what you mean, and that's a valid point.
I was thinking of maybe having two parameters: `max_pairs_per_bin` and `average_pairs_per_bin`.
The second would be used exactly as max_pairs_per_bin is used now. And `max_pairs_per_bin` would specify the upper limit of how many pairs can be added for a bin and a single inchikey.

Do you think that would be better?

In addition, we could also some function to evaluate how well certain parameters worked. Say, the bias remains to high, then this could be a sign to adapt those parameters or change the bins.

Reply:
Giving it a second thought I would prefer something closer to what you suggested @niekdejonge .
I would then switch to an `average_pairs_per_bin` and `max_oversampling_rate` (maybe better than `max_pairs_per_bin` because it could be kept at a default value, say of `max_oversampling_rate=2` and users are not obliged to change two parameters in accordance with each other).

We then take `max_oversampling_rate * max_pairs_per_bin` samples per bin.
At the very end we compute the total sum for each bin and compare it to `size * average_pairs_per_bin`. Should the sum be larger, then we'll remove pairs.

Reply:
I guess this is now covered with #146 .